### ISSUE TITLE
Error when upload file to Rocket Chat in subfolder

### ISSUE THREAD

**Issue by hoanpv3993** [RocketChat/Rocket.Chat#3073]
Your Rocket.Chat version: 0.28 (make sure you are running the latest)

Hi,
I have some problem with Rocket Chat in subfolder
- I can chat via web but can't access via App on Mobile
- When I upload any file, it can't display with Image, text,.. and I can't download it too.
Can you help me resolv them ?

This is error when i upload a photo.
![1](https://cloud.githubusercontent.com/assets/18733189/14909841/079f6c30-0e12-11e6-9466-23e5465fa5c8.png)

**Comment by marceloschmidt** [comment]
What are your file upload settings?

**Comment by dangkiena3** [comment]
I have same issue with images uploaded, please fix in next version. 

My rocket.chat version: 0.29

Thank you very much!

![image](https://cloud.githubusercontent.com/assets/18433986/15008060/b5722296-1207-11e6-9fbf-f0ebd6791ba9.png)

**Comment by hoanpv3993** [comment]
Sorry @marceloschmidt ,
I have had a holiday recently so I can't reply to you.
I don't not change any setting in Rocket Chat server. It is default  setting. 

![1](https://cloud.githubusercontent.com/assets/18733189/15009533/0ccf4ec0-1211-11e6-8bf9-76a47f63ce6b.png)

**Comment by dangkiena3** [comment]
This issue is file uploaded not run in sub directory like main page:

![image](https://cloud.githubusercontent.com/assets/18433986/15010072/66b8ac08-1214-11e6-89a3-16c79b48289b.png)

![image](https://cloud.githubusercontent.com/assets/18433986/15010077/6badcc0c-1214-11e6-9d4a-e2052ee837d5.png)

**Comment by dangkiena3** [comment]
Please help me :(

**Comment by geekgonecrazy** [comment]
@dangkiena3 using in a subdirectory is a very new addition, not many people are using it yet.  So we're still working out the kinks. Bare with us

**Comment by dangkiena3** [comment]
@geekgonecrazy many thank for reply. I'm still wait...

**Comment by dungvu-qe** [comment]
Same my issue, +1

**Comment by dangkiena3** [comment]
Now version 0.31 still does not fix this issue :(

**Comment by dangkiena3** [comment]
Version 0.32 still...

**Comment by dangkiena3** [comment]
Now is 0.33 version, please help me

**Comment by SQUIDwarrior** [comment]
Yup, having the same problem.

**Comment by SQUIDwarrior** [comment]
Is there a way I could hack this temporarily by hard-coding my subfolder somewhere?

**Comment by k0nsl** [comment]
@SQUIDwarrior:

Why not temporarily "solve" the issue by creating a simple rewrite rule? It should be quite easy depending on what web server you're running.

**Comment by SQUIDwarrior** [comment]
@k0nsl - No, that's a good workaround actually. I will give that a try! I totally didn't think of using a rewrite rule

**Comment by webwurst** [comment]
Still present with `v0.34`.

**Comment by AlecTroemel** [comment]
Im taking a stab at this..

**Comment by geekgonecrazy** [comment]
@AlecTroemel awesome, all yours! 👍  Let us know how it goes.  Feel free to open a PR if you need us to take a look at what you have.  PR's are great for discussing in context of the code :smile:

**Comment by mixrat** [comment]
same problem here :cry:

**Comment by AlecTroemel** [comment]
by changing https://github.com/AlecTroemel/Rocket.Chat/blob/develop/packages/rocketchat-message-attachments/client/messageAttachment.coffee#L11 to 
```javascript
return RocketChat.settings.get('Site_Url').replace(/\/$/, '') + url
```

rocketchat was able to find the image when being behind an nginx reverse proxy. However there is another (deeper) bug where it gives a 403 error on trying to get the image. this is because https://github.com/AlecTroemel/Rocket.Chat/blob/develop/lib/fileUpload.coffee#L16 is getting null for 

```javascript
Meteor._localStorage.getItem('Meteor.loginToken');
```

This can be worked around by setting *Administration --> File Upload --> Protect uploaded files* to false... though anyone can go and see those images..

**Comment by marceloschmidt** [comment]
@AlecTroemel what if you use `Meteor.absoluteUrl()` instead of `RocketChat.settings.get('Site_Url')`, would that solve it?

**Comment by AlecTroemel** [comment]
@marceloschmidt unfortunately (at least in my instance of rocketchat) `Meteor.absoluteUrl()` does not include the subdirectory. for example I've confirmed that the ROOT_URL env variable contains the full url with subdir `http://mydomain.com/subdir`, but Meteor.absoluteUrl() will only return `http://mydomain.com/`. Interestingly the ROOT_URL_PATH_PREFIX env var is undefined..which I don't think it should be

**Comment by AlecTroemel** [comment]
So it was my fault that ROOT_URL_PATH_PREFIX (I was trying to print it out before it was defined). From what I can tell, meteor takes the full url and splits it if there is the subdirectory. 

So http://mydomain.com/subdir leads to
```
absoluteUrl = http://mydomain.com/
__meteor_runtime_config__.ROOT_URL_PATH_PREFIX = /subdir
```


When the meteor account login sees that there is a `ROOT_URL_PATH_PREFIX` it adds some namespace information to the Meteor.loginToken localstorage variable name (code here https://github.com/meteor/meteor/blob/master/packages/accounts-base/localstorage_token.js#L100) 
so if there is a subdir, the loginToken key name changes, and this line will fail https://github.com/RocketChat/Rocket.Chat/blob/develop/lib/fileUpload.coffee#L16

heres an example of how those keys look for me
<img width="188" alt="screen shot 2016-07-05 at 4 12 39 pm" src="https://cloud.githubusercontent.com/assets/8620042/16600175/7d7c3a6c-42cb-11e6-9091-4bdf98282254.png">
I assume that ":/:/subdir" is the pattern, but I'm not 100% sure 

The following things solved the issue for me:
- checking for the path prefix, and concatenating the Meteor.loginToken key string with ":/:/subdir" 
- changing https://github.com/RocketChat/Rocket.Chat/blob/develop/packages/rocketchat-message-attachments/client/messageAttachment.coffee#L11 to 
```
return Meteor.absoluteUrl().replace(/\/$/, '') + __meteor_runtime_config__.ROOT_URL_PATH_PREFIX + url
```
I will make a PR so that others can review this.

**Comment by dangkiena3** [comment]
This issue was resolved! Thank you very much!

But I still can't use Mobile App with sub dir

**Comment by lunitic** [comment]
@AlecTroemel 

Is #4014 related to this issue.

My user gets the same issue but they we are not running in subdir, 
However we have a nginx infront of the RC.

**Comment by lunitic** [comment]
Was this fix in 0.37 released yesterday

**Comment by AlecTroemel** [comment]
@lunitic this fix was added in version 0.36

**Comment by mattlin** [comment]
I find this issue in 0.49.3 .   It is not fixed in  0.49.3  😢
